### DISCUSSION TITLE
Error System.Text.Json.JsonException

### DISCUSSION THREAD

**Discussion by faurepa** [pnp/pnpcore#1326] [Q&A] [answered]
Hello,
I am new a working with pnpcore.  I have downloaded the "pnpcore-dev" sample.  I have updated the Client and Tenant ID, and the location that I am trying to access in the appsettings.json file.

Independent of the location (tried two), I keep getting the following error.

System.Text.Json.JsonException
  HResult=0x80131500
  Message='<' is an invalid start of a value. Path: $ | LineNumber: 1 | BytePositionInLine: 0.
  Source=System.Text.Json
  StackTrace:
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at PnP.Core.Services.JsonMappingHelper.<MapJsonToModel>d__1.MoveNext()
   at PnP.Core.Services.BatchClient.<ProcessSharePointRestInteractiveResponse>d__61.MoveNext()
   at PnP.Core.Services.BatchClient.<ExecuteSharePointRestInteractiveAsync>d__60.MoveNext()
   at PnP.Core.Services.BatchClient.<ExecuteBatch>d__37.MoveNext()
   at PnP.Core.Model.BaseDataModel`1.<RequestAsync>d__96.MoveNext()
   at PnP.Core.Services.PnPContextFactory.<InitializeContextAsync>d__56.MoveNext()
   at PnP.Core.Services.PnPContextFactory.<CreateAsync>d__47.MoveNext()
   at PnP.Core.Services.PnPContextFactory.<CreateAsync>d__39.MoveNext()
   at PnP.Core.Services.PnPContextFactory.<CreateAsync>d__36.MoveNext()
   at Consumer.Program.<Main>d__0.MoveNext() in C:\Users\H173957\Documents\Code\pnpcore-dev\samples\Demo.Console\Program.cs:line 131
   at Consumer.Program.<Main>(String[] args)

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
JsonReaderException: '<' is an invalid start of a value. LineNumber: 1 | BytePositionInLine: 0.

I don't know if this is something with the site, or my code. 
I am not sure of where to start looking.  Any suggestions would be greatly appreciated.
Thank you

**Comment by faurepa** [comment]
I was wondering if there was a way to see what "using (var context = await pnpContextFactory.CreateAsync("DemoSite"))" line of code (where the failure occurs) is receiving as a webpage.  It looking for JSON, but may be getting a webpage.  All info line in the command window form the example indicate code 200, which should be good.  I had permssions errors (404 and 403), and those were resolved.  I jsut can't find what is receiving, or is there a way to change the format of the content that it should receive?

**Reply by jansenbe** [reply]
@faurepa : do you happen to have some kind of proxy in between you and SharePoint? Sometimes they mess up and you'll see this kind of errors. Doing a network trace (e.g. Fiddler) will show what really comes across the wire

**Reply by faurepa** [reply]
To the best of my knowledge, there is no proxy present in the system. I have installed Fiddler as suggested, I ran two scenarios:  once going to the site via my browser, and once using the PNP-core software. Now the issues is to find what I am looking for.

**Reply by jansenbe** [reply]
Look for the call to *.sharepoint.com which is failing (status code 4xx or 5xx) and then check the response body to see for any error information

**Reply by faurepa** [reply]
Hello, You were correct, there is an error, here is what the response body has.

```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Service</title>
   <style> ... </style>
  </head>
  <body>
    <div id="content">
      <p class="heading1">Service</p>
      <p xmlns="">There is no operation listening for <a href="https://usr192678-279.dprodmgd105.msft.net/_vti_bin/client.svc">https://usr192678-279.dprodmgd105.msft.net/_vti_bin/client.svc</a>, but there is an operation listening for <a href="https://honeywellprod.sharepoint.com/_vti_bin/client.svc/">https://honeywellprod.sharepoint.com/_vti_bin/client.svc/</a>, so you are being redirected there.</p>
    </div>
  </body>
</html>
```


I searched for "usr192678-279.dprodmgd105.msft.net", and it doesn't show up anywhere else in the fiddler report.
The error code for this action is 401, and the caching is listed as "private".  
Here is what was sent:
```
GET https://honeywellprod.sharepoint.com/teams/GlobalPolicyManual/Policies/Forms/AllItems.aspx/_vti_bin/client.svc HTTP/1.1
Host: honeywellprod.sharepoint.com
Authorization: Bearer
Access-Control-Allow-Origin: *
```

I hope that you might be able to provide some guidance as to what is going on.  Are there other API permissions required?
Thank you

**Reply by jansenbe** [reply]
@faurepa : that's a weird error. To simplify things, can you first try this minimalistic sample app and see if you've the same issue: https://pnp.github.io/pnpcore/demos/Demo.Console.Minimal/readme.html

**Reply by faurepa** [reply]
Hello,
I tried the minimal version, and I get an error 403.

info: System.Net.Http.HttpClient.SharePointRestClient.ClientHandler[100]
      Sending HTTP request GET https://honeywellprod.sharepoint.com/teams/GlobalPolicyManual/Policies/Forms/AllItems.aspx/_api/Web?$select=Id%2cUrl%2cRegionalSettings%2f*%2cRegionalSettings%2fDateFormat&$expand=RegionalSettings%2cRegionalSettings%2fTimeZone
info: System.Net.Http.HttpClient.SharePointRestClient.ClientHandler[101]
      Received HTTP response headers after 177.3786ms - 403
info: System.Net.Http.HttpClient.SharePointRestClient.LogicalHandler[101]
      End processing HTTP request after 217.9546ms - 403

Here are the details
PnP.Core.SharePointRestServiceException
  HResult=0x80131500
  Message=SharePoint Rest service exception
  Source=PnP.Core
  StackTrace:
   at PnP.Core.Services.BatchClient.<ExecuteSharePointRestInteractiveAsync>d__60.MoveNext()
   at PnP.Core.Services.BatchClient.<ExecuteBatch>d__37.MoveNext()
   at PnP.Core.Model.BaseDataModel`1.<RequestAsync>d__96.MoveNext()
   at PnP.Core.Services.PnPContextFactory.<InitializeContextAsync>d__56.MoveNext()
   at PnP.Core.Services.PnPContextFactory.<CreateAsync>d__47.MoveNext()
   at PnP.Core.Services.PnPContextFactory.<CreateAsync>d__42.MoveNext()
   at Program.<<Main>$>d__0.MoveNext() in C:\Users\H173957\Documents\Code\Demo.Console.Minimal\Program.cs:line 38

I noticed that I don't have these delegated permissions granted. 
Microsoft Graph -> Sites.Manage.All
SharePoint -> AllSites.Manage

I don't know if I am going to be able to get the Manage All permissions.  The software that I am writing is just going to read a file, Will that make a difference to the authentication?
here is a picture of what I have been granted permissions. I have requested for more, but I don't know if I am going to be able to get it.
![image](https://github.com/pnp/pnpcore/assets/26845996/a62ad1d6-a8bc-4da0-963a-f2128f500223)

**Reply by faurepa** [reply]
I was wondering why Sites.Manage.All and Allsites.Manage are required?  If I plan to just reading files.  I have no plans or even want to change files.

**Reply by jansenbe** [reply, marked as answer]
@faurepa : having the SharePoint AllSites.Read and Graph Sites.Read.All should be sufficient. Looking at the error I see "Policies/Forms/AllItems.aspx" in the URL. This makes me think you've specified the site collection wrongly. In your case the site collection to use would be https://honeywellprod.sharepoint.com/teams/GlobalPolicyManual. Can you cross check that?

**Reply by faurepa** [reply]
Using the Minimal software, I tried to use the path provided, still got the same error message.  asked IT to see if there is anything else that might be blocking the software.
I also tried using the demo program (not minimal), and I was able to get past the error message.
I commented out a bunch of code about teams, updating/adding/etc which I don't need, and it seems to be running correctly.  So it was just the siteUrl that was causing me an issue.

Now to figure out how to get a list of all the documents in the Policies folder.

**Reply by faurepa** [reply]
Still odd that "minimal" version doesn't work, but the full version does.  I appreciate your help, thank you

**Reply by faurepa** [reply]
I managed to get a full list of all folders in the root folder using the sample code provided.
Now to get the content of one specific folder.  I am essentially enumerating files using the sample code here:
https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#enumerating-files

```
await context.Web.LoadAsync(p => p.Folders);

var folder = context.Web.Folders.AsRequested().FirstOrDefault(p => p.Name == "Policies");

// Load files property of the folder
await folder.LoadAsync(p => p.Files);

foreach (var file in folder.Files.AsRequested())
{
    Console.WriteLine($"{file.Title} - Version: {file.Versions}");
}
```

Program fails on this line: await context.Web.LoadAsync(p => p.Folders);
Here is the console error:
info: System.Net.Http.HttpClient.SharePointRestClient.ClientHandler[100]
      Sending HTTP request GET https://honeywellprod.sharepoint.com/teams/GlobalPolicyManual/_api/web?$select=Id%2cFolders&$expand=Folders
info: System.Net.Http.HttpClient.SharePointRestClient.ClientHandler[101]
      Received HTTP response headers after 57.426ms - 403
info: System.Net.Http.HttpClient.SharePointRestClient.LogicalHandler[101]
      End processing HTTP request after 61.0094ms - 403

Is it possible that I am missing a permission?
Or have I not fully understood how this works?

**Reply by faurepa** [reply]
OK, thank you to your other support, I have found a solution:
```
//https://github.com/pnp/pnpcore/discussions/1322
var documentFolder = (await context.Web.Lists.GetByTitleAsync("Policies", p => p.RootFolder));
var allFiles = (await documentFolder.RootFolder.GetAsync(f => f.Files)).Files;

int i = 0;
foreach (var file in allFiles.AsRequested())
{
var extension = Path.GetExtension(file.Name);
Console.WriteLine($"{i} : Filename:{file.Name} - Path: {file.ServerRelativeUrl}");
i++;
}

I appreciate your help.
```
